packages/experimental-utils/src/ast-utils/predicates.ts

```diff
@@ -7,6 +7,26 @@ const isNodeOfType =
   ): node is TSESTree.Node & { type: NodeType } =>
     node?.type === nodeType;
 
+type ObjectEntry<BaseType> = [keyof BaseType, BaseType[keyof BaseType]];
+type ObjectEntries<BaseType> = Array<ObjectEntry<BaseType>>;
+const isNodeOfTypeWithConditions =
+  <
+    NodeType extends AST_NODE_TYPES,
+    Conditions extends Partial<TSESTree.Node & { type: NodeType }>,
+  >(
+    nodeType: NodeType,
+    conditions: Conditions,
+  ) =>
+  (
+    node: TSESTree.Node | null | undefined,
+  ): node is TSESTree.Node & { type: NodeType } & Conditions =>
+    node?.type === nodeType &&
+    (
+      Object.entries(conditions) as ObjectEntries<
+        TSESTree.Node & { type: NodeType }
+      >
+    ).every(([key, value]) => node[key] === value);
+
```

Comment:
On mobile so this will probs look garbage but it'll get my request across (hopefully)

```suggestion
type ObjectEntry<BaseType> = [keyof BaseType, BaseType[keyof BaseType]];
type ObjectEntries<BaseType> = Array<ObjectEntry<BaseType>>;
const isNodeOfTypeWithConditions =
  <
    NodeType extends AST_NODE_TYPES,
    Conditions extends Partial<TSESTree.Node & { type: NodeType }>,
  >(
    nodeType: NodeType,
    conditions: Conditions,
  ) => {
    const entries = Object.entries(conditions) as ObjectEntries<
        TSESTree.Node & { type: NodeType }
      >;
    return (
      node: TSESTree.Node | null | undefined,
    ): node is TSESTree.Node & { type: NodeType } & Conditions =>
      node?.type === nodeType &&
      (
        Object.entries(conditions)
      ).every(([key, value]) => node[key] === value);

```

Reply:
I see what you meant now.

Can you tell me what the main difference is on a performance level (or any other level)?
I don't see any real difference except having the extra variable 🤔

Reply:
current version calculates `Object.entries` on every call to the function.
this suggested version calculates `Object.entries` once and only once (on function declaration).

Just saves some useless repeated recalculation! (perf wise about 10% faster)

I also realised I cocked it up in my mobile suggestion - it should be:
```ts
const isNodeOfTypeWithConditions =
  <
    NodeType extends AST_NODE_TYPES,
    Conditions extends Partial<TSESTree.Node & { type: NodeType }>,
  >(
    nodeType: NodeType,
    conditions: Conditions,
  ) => {
    const entries = Object.entries(conditions) as ObjectEntries<
      TSESTree.Node & { type: NodeType }
    >;
    return (
      node: TSESTree.Node | null | undefined,
    ): node is TSESTree.Node & { type: NodeType } & Conditions =>
      node?.type === nodeType &&
      entries.every(([key, value]) => node[key] === value);
  };
```